src/protocols/secure_channel/CASESession.cpp

```diff
@@ -1173,40 +1413,68 @@ CHIP_ERROR CASESession::ValidateReceivedMessage(ExchangeContext * ec, const Payl
     }
 
     VerifyOrReturnError(!msg.IsNull(), CHIP_ERROR_INVALID_ARGUMENT);
-    VerifyOrReturnError(payloadHeader.HasMessageType(mNextExpectedMsg) ||
-                            payloadHeader.HasMessageType(Protocols::SecureChannel::MsgType::CASE_SigmaErr),
-                        CHIP_ERROR_INVALID_MESSAGE_TYPE);
-
     return CHIP_NO_ERROR;
 }
 
 CHIP_ERROR CASESession::OnMessageReceived(ExchangeContext * ec, const PayloadHeader & payloadHeader,
                                           System::PacketBufferHandle && msg)
 {
-    CHIP_ERROR err = ValidateReceivedMessage(ec, payloadHeader, msg);
+    CHIP_ERROR err                            = ValidateReceivedMessage(ec, payloadHeader, msg);
+    Protocols::SecureChannel::MsgType msgType = static_cast<Protocols::SecureChannel::MsgType>(payloadHeader.GetMessageType());
     SuccessOrExit(err);
 
-    switch (static_cast<Protocols::SecureChannel::MsgType>(payloadHeader.GetMessageType()))
+    err = CHIP_ERROR_INVALID_MESSAGE_TYPE;
+
+    switch (mState)
     {
-    case Protocols::SecureChannel::MsgType::CASE_SigmaR1:
-        err = HandleSigmaR1_and_SendSigmaR2(std::move(msg));
+    case kInitialized:
+        if (msgType == Protocols::SecureChannel::MsgType::CASE_Sigma1)
+        {
+            err = HandleSigma1_and_SendSigma2(std::move(msg));
+        }
         break;
+    case kSentSigma1:
+        switch (static_cast<Protocols::SecureChannel::MsgType>(payloadHeader.GetMessageType()))
+        {
+        case Protocols::SecureChannel::MsgType::CASE_Sigma2:
+            err = HandleSigma2_and_SendSigma3(std::move(msg));
+            break;
 
-    case Protocols::SecureChannel::MsgType::CASE_SigmaR2:
-        err = HandleSigmaR2_and_SendSigmaR3(std::move(msg));
-        break;
+        case Protocols::SecureChannel::MsgType::CASE_Sigma2Resume:
+            err = HandleSigma2Resume(std::move(msg));
+            break;
 
-    case Protocols::SecureChannel::MsgType::CASE_SigmaR3:
-        err = HandleSigmaR3(std::move(msg));
-        break;
+        case MsgType::StatusReport:
+            err = HandleStatusReport(std::move(msg));
+            break;
 
-    case Protocols::SecureChannel::MsgType::CASE_SigmaErr:
-        err = HandleErrorMsg(std::move(msg));
+        default:
+            break;
+        };
         break;
+    case kSentSigma2:
+        switch (static_cast<Protocols::SecureChannel::MsgType>(payloadHeader.GetMessageType()))
+        {
+        case Protocols::SecureChannel::MsgType::CASE_Sigma3:
+            err = HandleSigma3(std::move(msg));
+            break;
+
+        case MsgType::StatusReport:
+            err = HandleStatusReport(std::move(msg));
+            break;
 
+        default:
+            break;
+        };
+        break;
+    case kSentSigma3:
+    case kSentSigma2Resume:
+        if (msgType == Protocols::SecureChannel::MsgType::StatusReport)
+        {
+            err = HandleStatusReport(std::move(msg));
+        }
+        break;
     default:
```

Comment:
Add a comment that the default case fall through to having the default `err` value indicating message not handled. Please also log the message type and the fact an error occurred.